### DISCUSSION TITLE
Outputing a2dp_sink to a FilteredStream

### DISCUSSION THREAD

**Discussion by aqw42** [pschatzmann/ESP32-A2DP#620] [Q&A] [answered]
Hi there !
I'm trying to create this pipeline : 

BluetoothA2dpSink -> FilteredStream -> AnalogAudioStream

(a2dp is set in mono mode, so only one channel is needed)

And i can't get it working !
I tried manually creating an a2dp data callback, I tried with copiers, but something is missing.
I don't understand how to link these together while keeping the configuration the same along the pipeline (for example, how to have the same rate)
I'd like to know what do you think is the **proper** way to do that.

I'm using Arduino IDE with M5StickC (ESP32 PICO 4M)

Thanks for your help,
and thanks for this library that is overall very helpful (and keeped up to date with the latest esp-idf DAC API updates ahah)

**Comment by pschatzmann** [comment, marked as answer]
Do you really want to use a FilteredStream ? I think the [Equilizer3Bands](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_equilizer3_bands.html) is much easier to use.

I am confused about your mono statement: I have always seen stereo so far. Maybe you want to convert the [Stereo into Mono](https://github.com/pschatzmann/arduino-audio-tools/wiki/Converting-the-Data-Format) ?

There are always many ways to do the same thing:

- You could use the A2DPStream from the AudioTools to build your chain
- You can use the BluetoothA2dpSink and in the callback you just write the data to your FilteredStream

And please condider the [advice given in the Wiki](https://github.com/pschatzmann/arduino-audio-tools/wiki/It's-not-working): build your solution in multiple steps so that you can test them individually.

**Reply by aqw42** [reply]
Thanks for your fast answer !
I made a quick test using the Equilizer3Bands in stereo, it works like a charm (with a lot of noise but i guess that can be explained by my bread board and a 8-bit internal DAC)

Here is my code if someone else wants to achieve the same : 
```
#include "AudioTools.h"
#include "BluetoothA2DPSink.h"

AnalogAudioStream out;
Equilizer3Bands eq(out);
BluetoothA2DPSink a2dp_sink;

void a2dp_cb(const uint8_t *data, uint32_t len)
{
  eq.write(data, len);
}

void setup() {
  auto config = out.defaultConfig();
  config.channels = 2;
  config.bits_per_sample = 16;
  config.sample_rate = 44100;
  out.begin(config);

  auto &cfg = eq.defaultConfig();
  cfg.gain_low = 1;
  cfg.freq_low = 100;
  cfg.gain_medium = 0;
  cfg.gain_high = 0;
  cfg.channels = 2;
  cfg.bits_per_sample = 16;
  cfg.sample_rate = 44100;
  eq.begin(cfg);
  
  a2dp_sink.set_stream_reader(a2dp_cb, false);
  a2dp_sink.start("Jean paul", true);
}


void loop() {
}
```

Do you see anything wrong in this code ?
Once again thanks a lot, I was struggling on this and you solved it in minutes

**Comment by harryberlin** [comment]
have a look at this:
https://github.com/pschatzmann/ESP32-A2DP/issues/271#issuecomment-2163688278
